accepted/0042-isolated-mode.md

```diff
@@ -0,0 +1,347 @@
+
+## Motivation
+
+The introduction of workspaces in npm v7 brought the ability to split large code-bases into smaller pieces of code with declared dependencies between them.
+
+Several build tools took advantage of these declared dependencies to split the builds into multiple steps (one per workspace). These build tools take advantage of the declared dependencies between workspaces to only rebuild the workspaces that have not changed since the last build. With the current way npm works, any dependency change to any workspace may affect any other workspace. This means that the build systems need to rebuild the entire project every time a single workspace modifies a dependency. This is inefficient and does not scale.
+
+## Naming
+
+This mode is called `isolated-mode` because it makes the boundaries between packages and workspaces stronger, making them more isolated. 
+
+
+To make the discussions more efficient, we could also name the current default installation mode of npm. We will call this mode `hoisted-mode` as packages are shared across workspaces by hoisting them to the root of the project. This term is already known by developers familiar with yarn or pnpm.
+
+## More on the problem
+
+Packages and workspaces declare their dependencies to npm by using special fields in their `package.json`. The Node.js runtime is unaware of these special fields, instead it uses a [module resolution algorithm](https://nodejs.org/api/modules.html) to determine the meaning of a module importing another module. npm bridges this gap by converting the declared dependency graph into a folder structure which makes sense to Node.js. This operation is call `reification`.
+
+When converting the dependency graph to a folder structure, `hoisted-mode` loses information. Multiple different dependency graphs can be converted to an identical folder structure. Since the folder structure is all that Node.js uses to resolve modules, Node.js misses some information about dependencies. This loss of information is most frequently visible through the fact that workspaces are able to successfully import the dependencies declared by other workspaces.
+
+### Forgetting to declare dependencies
+
+When workspaces can successfully use code of a package without having a dependency on it, people forget to declare their dependencies. This leads to situations where updating the dependencies of one workspace breaks a seemingly unrelated workspace.
+
+It is worth noting that static code analysis tools can help significantly reduce the frequency of these mistakes. While static code analysis tools are probably sufficient for small projects, they are not for large projects. "isolated-mode" not only helps get the dependencies right but also _proves_ that the dependencies are correct. This proof is essential to build trustable build systems which use caching, scoped installations.
+
+### Duplication
```

Comment:
Is this something orthogonal? iow, this seems like something worth solving for both modes, not just for this one.

Reply:
@ljharb can you clarify what would be worth solving for both modes?

Reply:
In the default mode, could you not still use symlinking to share package versions across projects and save disk space?

Reply:
It is difficult to use symlinks without fully implementing a store like proposed in this RFC. The reason is that when you require() or require.resolve(), you always get back a real path, so you look the context of which you come from (unless you have that context encoded in the store like this RFC suggests).

Another option for using symlinks is to use the preserve symlink node option, but it comes with bad implications like bad perf because a module can be instanciated many times.
 https://nodejs.org/api/cli.html#cli_preserve_symlinks

Reply:
We do use symlinks in some cases to resolve otherwise impossible scenarios.  For example, in the "pathological nesting loop" scenario, we use symlinks to avoid an infinite regress.  The simplest formulation of that scenario is:

```
x@1 -> (y@1)
x@2 -> (y@2)
y@1 -> (x@2)
y@2 -> (x@1)
```

When reifying, the first time we notice that a package is descended from a package of the same name that does not match the dep being installed, which is in turn descended from a package of the same name which _does_ match the dep being installed, we place a symbolic link instead of unpacking it a second time.

So instead of this:

```
root
+-- x@1
+-- y@1
    +-- x@2
    +-- y@2
        +-- x@1
        +-- y@1
            +-- x@2 ...
```

we have this instead:

```
root
+-- x@1
+-- y@1
    +-- x@2
    +-- y@2
        +-- x@1
        +-- y -> root/node_modules/y
```

I have explored a "dedupe harder" mode.  In this mode, we walk the nodes in the tree, and any that match another instance in the tree are replaced with a symbolic link to the other copy.  It's not very complicated code to deal with, but it doesn't save much in disk space or install time in the experiments we ran, and it resulted in a package tree with symbolic links that were somewhat haphazard for a human who might go poking around in node_modules.

It may be worth exploring again in the context of a massive monorepo, if we want to get the benefits of aggressive deduplication _without_ dependency isolation.

For that reason, I'd suggest the opt-in here be a `--mode=(default|isolated|...)` rather than a boolean, so that we can expand the set of reification modes in the future if we choose.

Reply:
In fact, we're likely to eventually provide a `--mode=importmap` reification strategy, which would potentially deliver the value of `--mode=isolated`, but without any symlinks at all.

Reply:
I added some bikeshedding comments on this.